x/rewards/ibc_middleware.go

```diff
@@ -0,0 +1,169 @@
+	counterparty channeltypes.Counterparty,
+	counterpartyVersion string,
+) (version string, err error) {
+	return im.app.OnChanOpenTry(ctx, order, connectionHops, portID, channelID, chanCap, counterparty, counterpartyVersion)
+}
+
+func (im IBCMiddleware) OnChanOpenAck(
+	ctx sdk.Context,
+	portID, channelID string,
+	counterpartyChannelID string,
+	counterpartyVersion string,
+) error {
+	return im.app.OnChanOpenAck(ctx, portID, channelID, counterpartyChannelID, counterpartyVersion)
+}
+
+func (im IBCMiddleware) OnChanOpenConfirm(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+func (im IBCMiddleware) OnChanCloseInit(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+func (im IBCMiddleware) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket checks the packet's memo and funds the IPRPC pool accordingly. If the memo is not the expected JSON,
+// the packet is transferred normally to the next IBC module in the transfer stack
+func (im IBCMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) exported.Acknowledgement {
+	// unmarshal the packet's data with the transfer module codec (expect an ibc-transfer packet)
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// extract the packet's memo
+	memo, err := im.keeper.ExtractIprpcMemoFromPacket(ctx, data)
+	if errors.Is(err, types.ErrMemoNotIprpcOverIbc) {
+		// not a packet that should be handled as IPRPC over IBC (not considered as error)
+		utils.LavaFormatDebug("rewards module IBC middleware processing skipped, memo is invalid for IPRPC over IBC funding",
+			utils.LogAttr("memo", memo),
+		)
+		return im.app.OnRecvPacket(ctx, packet, relayer)
+	} else if errors.Is(err, types.ErrIprpcMemoInvalid) {
+		// memo is in the right format of IPRPC over IBC but the data is invalid
+		utils.LavaFormatWarning("rewards module IBC middleware processing failed, memo data is invalid", err,
+			utils.LogAttr("memo", memo))
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// change the ibc-transfer packet receiver address to be the rewards module address and empty the memo
+	data.Receiver = im.keeper.GetModuleAddress()
+	data.Memo = ""
+	marshelledData, err := transfertypes.ModuleCdc.MarshalJSON(&data)
+	if err != nil {
+		utils.LavaFormatError("rewards module IBC middleware processing failed, cannot marshal packet data", err,
+			utils.LogAttr("data", data))
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+	packet.Data = marshelledData
+
+	// call the next OnRecvPacket() of the transfer stack to make the rewards module get the IBC tokens
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
```

Comment:
add an comment why ack == nil is bad

Reply:
done. see https://github.com/lavanet/lava/commit/c5b5020fc7184d9099d5fe578c69a2ba43db320e